### DISCUSSION TITLE
Limit subseconds precision

### DISCUSSION THREAD

**Discussion by fulmicoton** [time-rs/time#418] [Question]
I am trying to use time to format UTC into rfc3339.
Unfortunately the nanoseconds precision is a bit too high and appears very verbose in the logs.
I would love to either have a way to limit the precision of a format to milliseconds precision,
or at least make it possible to define it using a custom format.

With time's DSL, I don't think I can express `"%Y-%m-%dT%H:%M:%S%.3f"`

**Comment by jhpratt** [comment]
```rust
format_description!("[year]-[month]-[day]T[hour]:[minute]:[second].[subsecond digits:3]")
```

I'm assuming you need something _like_ RFC3339, not RFC3339 itself, given that you mentioned the possibility of custom formats.

**Reply by fulmicoton** [reply]
Yes. I couldn't find the digits:3 modifier in the documentation. I confirm this solves my problem! Thank you

**Reply by fulmicoton** [reply]
Oh and I would rather use RFC3339 itself, but it is not possible to limit the digit precision there today, is it?

**Reply by jhpratt** [reply]
No, that's not possible. It's also not planned, as well-known formats comply strictly with the relevant specification.

**Reply by fulmicoton** [reply]
rfc3339 does not mention anything about the precision of `time-secfrac` though.

**Reply by jhpratt** [reply]
It indicates one or more digits. When parsing this is done by consuming _all_ digits (even if those after the ninth are functionally ignored). When formatting the logical thing to do is to produce as many digits as possible.

**Reply by fulmicoton** [reply]
Understood. I'll rely on the format_description macro then. Thank you!
 
In case this comes up again, my use case is logging with tokio-tracing, and 9 digits precision is hurting the readability of the logs.

**Reply by andrewhickman** [reply]
Would you be open to making this configurable through an enum like https://docs.rs/chrono/latest/chrono/enum.SecondsFormat.html?

Perhaps we could add a new well-known time format like `Rfc3339WithOptions { ... }`?

My own use case is implementing serialization of the protobuf `google.protobuf.Timestamp` type. The spec requires emitting 3, 6 or 9 fractional digits depending on the value.

**Reply by jhpratt** [reply]
The thing is, "RFC3339 with options" is not RFC3339, so even including the name is misleading. It's already possible to set the number of digits, as was mentioned earlier in this discussion. [`SubsecondDigits`](https://time-rs.github.io/api/time/format_description/modifier/enum.SubsecondDigits.html) exists if you need to do something manually.
